### PR TITLE
Add aws_transfer_server table

### PR DESCRIPTION
**Pull request by jramosf**
This PR adds a new table `aws_transfer_server` for the service AWS Transfer Family

# Example query results
<details>
  <summary>Results</summary>

```
select
  server_id,
  domain,
  identity_provider_type,
  endpoint_type
from
  aws_transfer_server;
+---------------------+--------+------------------------+---------------+
| server_id           | domain | identity_provider_type | endpoint_type |
+---------------------+--------+------------------------+---------------+
| s-7c591529ce4d5064a | S3     | SERVICE_MANAGED        | VPC           |
| s-bf24bfc2a95d33bcb | S3     | SERVICE_MANAGED        | VPC           |
| s-23d72a1217988f65f | S3     | SERVICE_MANAGED        | VPC           |
| s-099baa3a90c8e932a | S3     | SERVICE_MANAGED        | VPC           |
+---------------------+--------+------------------------+---------------+


select
  server_id,
  domain,
  identity_provider_type,
  endpoint_type,
from
  aws_transfer_server
where
  state = 'OFFLINE';
+---------------------+--------+------------------------+---------------+---------+
| server_id           | domain | identity_provider_type | endpoint_type | state   |
+---------------------+--------+------------------------+---------------+---------+
| s-099baa3a90c8e932a | S3     | SERVICE_MANAGED        | VPC           | OFFLINE |
+---------------------+--------+------------------------+---------------+---------+

select
  server_id,
  user_count
from
  aws_transfer_server
order by
  user_count desc;
+---------------------+------------+
| server_id           | user_count |
+---------------------+------------+
| s-7c591529ce4d5064a | 744        |
| s-0995670b38116cb9b | 696        |
| s-099baa3a90c8e932a | 29         |
| s-23d72a1217988f65f | 1          |
+---------------------+------------+

select
  server_id,
  domain,
  identity_provider_type,
  endpoint_type,
  workflow_details ->> 'OnUpload' as on_upload_workflow
from
  aws_transfer_server
limit 1;
+---------------------+--------+------------------------+---------------+---------------------------------------------------------------------------------------------------------------------------------------------------------+
| server_id           | domain | identity_provider_type | endpoint_type | workflow_details                                                                                                                                        |
+---------------------+--------+------------------------+---------------+---------------------------------------------------------------------------------------------------------------------------------------------------------+
| s-1814e0ed169759723 | S3     | SERVICE_MANAGED        | VPC           | [{"ExecutionRole":"arn:aws:iam::012345678901:role/my-workflow-dev","WorkflowId":"w-3d4a97eb2677351ff"}] |
+---------------------+--------+------------------------+---------------+---------------------------------------------------------------------------------------------------------------------------------------------------------+

select
  server_id,
  domain,
  identity_provider_type,
  endpoint_type,
  structured_log_destinations
from
  aws_transfer_server
where structured_log_destinations is not null;
+---------------------+--------+------------------------+---------------+---------------------------------------------------------------------------------------+
| server_id           | domain | identity_provider_type | endpoint_type | structured_log_destinations                                                           |
+---------------------+--------+------------------------+---------------+---------------------------------------------------------------------------------------+
| s-1814e0ed169759723 | S3     | SERVICE_MANAGED        | VPC           | ["arn:aws:logs:us-west-2:012345678901:log-group:/aws/transfer/s-1814e0ed169759723:*"] |
+---------------------+--------+------------------------+---------------+---------------------------------------------------------------------------------------+
```
</details>

### CODE REVIEW

**Review by raffimohammed — APPROVED** [review]

**Comment by jramosf** [comment]
@misraved Any chance this gets merged? Thanks!

**Comment by jramosf** [comment]
Responding inline

> Thank you so much @jramosf for the PR 👍 , great work 🚀 !!
> 
> The overall structure of the table and the example queries look good. A few questions and suggestions before we merge in this PR:
> 
> * Could you please add 1-2 more example queries in the table doc? JSON columns like `workflow_details` and `structured_log_destinations` could be used as independent queries.

Added queries to the markdown doc in [095d7cd](https://github.com/turbot/steampipe-plugin-aws/pull/1909/commits/095d7cd7804c9a4c21006dd42a95ba20d2148b05), as well as example results in this Pull Request description.
> * The changes made in https://github.com/turbot/steampipe-plugin-aws/pull/1909/files#diff-e4f9949b3a358bc0195d3fe1e10d6a40039e48e37d6a1051fcc30a72dc67cdb0R1689-R1696 look good, did you get a chance to test them?

I've run the tests which went fine, and actually I've also been using the modified binary from this build as regular basis.
> * This PR updates the AWS Go SDK, did you come across any breaking changes? I took a look at https://github.com/turbot/steampipe-plugin-aws/pull/1909/files#diff-e4f9949b3a358bc0195d3fe1e10d6a40039e48e37d6a1051fcc30a72dc67cdb0R1689-R1696 and couldn't find any, but I just wanted to make sure that I did not miss any breaking change.

I checked the changelog (the increase is not a major) and didn't find any breaking change. This is expected from the SDK Semantic Versioning.

Cheers,

**Comment by misraved** [comment]
Thanks @jramosf for the answers 👍.

**Comment by jramosf** [comment]
Hi @misraved / @ParthaI , is this good to merge? Cheers

File: aws/table_aws_transfer_server.go [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+				Name:        "user_count",
+				Description: "Specifies the number of users that are assigned to a server.",
+				Type:        proto.ColumnType_INT,
+			},
+			{
+				Name:        "certificate",
+				Description: "Specifies the ARN of the Amazon Web ServicesCertificate Manager (ACM) certificate.",
+				Type:        proto.ColumnType_STRING,
+				Hydrate:     getTransferServer,
+				Transform:   transform.FromField("Certificate"),
```
Comment (ParthaI) [inline aws/table_aws_transfer_server.go:76]: @jramosf, do we require the transform function here? I believe the default [transform function](https://github.com/turbot/steampipe-plugin-aws/blob/main/aws/plugin.go#L20) will handle that conversion for us (from snake case to Upper Camel Case).

Comment (ParthaI) [reply aws/table_aws_transfer_server.go:76]: Please apply the same update to other columns wherever it is applicable.

File: aws/table_aws_transfer_server.go [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+				Name:        "title",
+				Description: resourceInterfaceDescription("title"),
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromField("ServerId"),
+			},
+			{
+				Name:        "akas",
+				Description: resourceInterfaceDescription("akas"),
+				Type:        proto.ColumnType_JSON,
+				Transform:   transform.FromField("Arn").Transform(arnToAkas),
```
Comment (ParthaI) [inline aws/table_aws_transfer_server.go:166]:
```suggestion
				Transform:   transform.FromField("Arn").Transform(transform.EnsureStringArray),
```

File: aws/table_aws_transfer_server.go [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+	svc, err := TransferClient(ctx, d)
+	if err != nil {
+		plugin.Logger(ctx).Error("aws_transfer_server.listTransferServers", "client_error", err)
+		return nil, err
+	}
+	if svc == nil {
+		return nil, nil
+	}
+	// Limiting the results
+	maxLimit := int32(100)
```
Comment (ParthaI) [inline aws/table_aws_transfer_server.go:185]: I would like to request an update to the default value of `maxLimit` to `1000`, in accordance with the [documentation](https://docs.aws.amazon.com/transfer/latest/userguide/API_ListServers.html). Could you please make this change?

File: aws/table_aws_transfer_server.go [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+	if d.QueryContext.Limit != nil {
+		limit := int32(*d.QueryContext.Limit)
+		if limit < maxLimit {
+			if limit < 1 {
+				maxLimit = 1
+			} else {
+				maxLimit = limit
+			}
```
Comment (ParthaI) [inline aws/table_aws_transfer_server.go:193]:
```suggestion
			maxLimit = limit
```
If a user passes a limit value less than one, the Steampipe SDK handles it automatically, and there's no need for us to handle it explicitly.

File: aws/table_aws_transfer_server.go [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+		return nil, err
+	}
+
+	var serverID string
+	if h.Item != nil {
+		serverID = *h.Item.(types.ListedServer).ServerId
+	} else {
+		serverID = d.EqualsQuals["server_id"].GetStringValue()
+	}
+
```
Comment (ParthaI) [inline aws/table_aws_transfer_server.go:243]: Could you please add an empty check for `serverID`?
We should enforce restrictions on API calls when a user provides an empty string in the `server_id` WHERE clause, such as (`SELECT * FROM aws_transfer_server WHERE server_id = ''`).

File: aws/table_aws_transfer_server.go [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+//// TABLE DEFINITION
+
+func tableAwsTransferServer(_ context.Context) *plugin.Table {
+	return &plugin.Table{
+		Name:        "aws_transfer_server",
+		Description: "AWS Transfer Server",
+		Get: &plugin.GetConfig{
+			KeyColumns: plugin.SingleColumn("server_id"),
+			Hydrate:    getTransferServer,
```
Comment (ParthaI) [inline aws/table_aws_transfer_server.go:24]: Please handle the not found error code([Add ignore config](https://github.com/turbot/steampipe-plugin-aws/blob/main/aws/table_aws_acm_certificate.go#L30)) for the API call. The `DescribeService` API throws [ResourceNotFoundException](https://docs.aws.amazon.com/transfer/latest/userguide/API_DescribeServer.html) error when a resource is not found by the AWSTransfer Family service.

**Review by ParthaI — COMMENTED** [review]
@jramosf, I left a few review comments. Could you please take a look, thanks!

Additionally, 
- Please include a test case for the resource not found in the integration.
- Could we also add an additional example of a join query with other tables? (e.g: From the table column structure I can see we can join this table with `aws_acm_certificate`)

**Pull request merged**